crates/sui-graphql-config/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+    NestedMeta,
+};
+
+/// Attribute macro to be applied to config-based structs. It ensures that the struct derives serde
+/// traits, and `Debug`, that all fields are renamed with "kebab case", and adds a `#[serde(default
+/// = ...)]` implementation for each field that ensures that if the field is not present during
+/// deserialization, it is replaced with its default value, from the `Default` implementation for
+/// the config struct.
+#[allow(non_snake_case)]
+#[proc_macro_attribute]
+pub fn GraphQLConfig(_attr: TokenStream, input: TokenStream) -> TokenStream {
+    let DeriveInput {
+        attrs,
+        vis,
+        ident,
+        generics,
+        data,
+    } = parse_macro_input!(input as DeriveInput);
+
+    let Data::Struct(DataStruct {
+        struct_token,
+        fields,
+        semi_token,
+    }) = data
+    else {
+        panic!("GraphQL configs must be structs.");
+    };
+
+    let Fields::Named(FieldsNamed {
+        brace_token: _,
+        named,
+    }) = fields
+    else {
+        panic!("GraphQL configs must have named fields.");
+    };
+
+    // Figure out which derives need to be added to meet the criteria of a config struct.
+    let extra_derives = extra_derives(&attrs);
+
+    // Extract field names once to avoid having to check for their existence multiple times.
+    let fields_with_names: Vec<_> = named
+        .iter()
+        .map(|field| {
+            let Some(ident) = &field.ident else {
+                panic!("All fields must have an identifier.");
+            };
+
+            (ident, field)
+        })
+        .collect();
+
+    // Generate the fields with the `#[serde(default = ...)]` attribute.
+    let fields = fields_with_names.iter().map(|(name, field)| {
+        let default = format!("{ident}::__default_{name}");
+        quote! { #[serde(default = #default)] #field }
+    });
+
+    // Generate the default implementations for each field.
+    let defaults = fields_with_names.iter().map(|(name, field)| {
+        let ty = &field.ty;
+        let fn_name = format_ident!("__default_{}", name);
+        let cfg = extract_cfg(&field.attrs);
+
+        quote! {
+            #[doc(hidden)] #cfg
+            fn #fn_name() -> #ty {
+                Self::default().#name
+            }
+        }
+    });
+
+    TokenStream::from(quote! {
+        #[derive(#(#extra_derives),*)]
+        #[serde(rename_all = "kebab-case")]
+        #(#attrs)* #vis #struct_token #ident #generics {
+            #(#fields),*
+        } #semi_token
+
+        impl #ident {
+            #(#defaults)*
+        }
+    })
+}
+
+/// Return a set of derives that should be added to the struct to make sure it derives all the
+/// things we expect from a config, namely `Serialize`, `Deserialize`, and `Debug`.
+///
+/// We cannot add extra derives unconditionally, because they will conflict with existing ones.
+fn extra_derives(attrs: &[Attribute]) -> BTreeSet<Ident> {
```

Comment:
maybe `core_derives` instead, since this constitutes the set of derives we do want to apply to a struct so it conforms to a "config"